### PR TITLE
Compute projections

### LINKED ISSUE #209: Projection for actuals
Adapt calculation and test calculation of projections for Actuals.
Advance and Overdue Actuals, can be entirely released. 

Allow the calculation of Actuals (written) as cashflow release for the current period. 
This should happen only when Actuals are not imported yet. 
Actuals for CF amount types that don't have a ACT value should also be deleted upon ACT import.

### LINKED ISSUE #210: Projection calculation of complex report (CSM family)
Compute projection of CSM family report and DAC

### PR DESCRIPTION
**Pull request by dcolleoni**

### CODE REVIEW

File: ifrs17-template/Test/Data/Actuals_CH_2020_12_MTUP10pct.csv
Comment (dcolleoni) [inline ifrs17-template/Test/Data/Actuals_CH_2020_12_MTUP10pct.csv:1]: Reset

File: ifrs17/Import/2ImportScope-PresentValue.ipynb
Comment (dcolleoni) [inline ifrs17/Import/2ImportScope-PresentValue.ipynb:1]: Add documentation

File: ifrs17/DataModel/DataStructure.ipynb
Comment (dcolleoni) [inline ifrs17/DataModel/DataStructure.ipynb:1]: Clean up

File: ifrs17/Import/3ImportScope-Actuals.ipynb [hunk 1]
```diff
@@ -175,17 +203,38 @@
...
-                "\n                                    .SelectMany(at => accidentYears",
-                "\n                                        .Select(ay => GetScope<ActualBase>((Identity, at, EstimateType, ay)))).ToArray();",
+                "public interface AccrualActual : IScope<(ImportIdentity Id, string AmountType, string EstimateType, int? AccidentYear), ImportStorage>{",
+                "\n    static ApplicabilityBuilder ScopeApplicabilityBuilder(ApplicabilityBuilder builder) =>",
+                "\n        builder.ForScope<AccrualActual>(s => s",
+                "\n            .WithApplicability<AccrualEmptyValues>(x => !x.GetStorage().GetAllAocSteps(StructureType.AocAccrual).Contains(x.Identity.Id.AocStep))",
+                "\n            .WithApplicability<AccrualEndOfPeriod>(x => !x.GetStorage().IsSecondaryScope(x.Identity.Id.DataNode) && x.Identity.Id.AocType == AocTypes.EOP)",
+                "\n            .WithApplicability<AccrualEmptyValues>(x => !x.GetStorage().IsSecondaryScope(x.Identity.Id.DataNode) && x.GetStorage().GetShift(x.Identity.Id.ProjectionPeriod) > MonthInAYear) // projections beyond ReportingYear +1",
```
Comment (dcolleoni) [inline ifrs17/Import/3ImportScope-Actuals.ipynb]: All these conditions can be rewritten in terms of NextYearProjection

File: ifrs17/Import/4ImportScope-TechnicalMargin.ipynb [hunk 2]
```diff
@@ -539,7 +545,7 @@
                 "\n    static ApplicabilityBuilder ScopeApplicabilityBuilder(ApplicabilityBuilder builder) => ",
                 "\n        builder.ForScope<AllocateTechnicalMargin>(s => s",
                 "\n                                             .WithApplicability<ComputeAllocateTechnicalMarginWithIfrsVariable>(x => x.GetStorage().IsSecondaryScope(x.Identity.DataNode))",
-                "\n                                             .WithApplicability<AllocateTechnicalMarginForBopProjection>(x => x.Identity.AocType == AocTypes.BOP && x.Identity.Novelty == Novelties.I && x.Identity.ProjectionPeriod > 0)",
+                "\n                                             //.WithApplicability<AllocateTechnicalMarginForBopProjection>(x => x.Identity.AocType == AocTypes.BOP && x.Identity.Novelty == Novelties.I && x.Identity.ProjectionPeriod > 0)",
```
Comment (dcolleoni) [inline ifrs17/Import/4ImportScope-TechnicalMargin.ipynb]: clean up

File: ifrs17/Import/6ImportScope-Compute.ipynb [hunk 3]
```diff
@@ -164,10 +166,92 @@
...
+                "\n}"
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "//TODO check the exact result of importing only Opening",
```
Comment (dcolleoni) [inline ifrs17/Import/6ImportScope-Compute.ipynb]: Follow up on these todos

File: ifrs17/Import/ImportStorage.ipynb [hunk 4]
```diff
@@ -353,8 +356,8 @@
...
                 "\n    public IEnumerable<IfrsVariable> GetIfrsVariables(string dataNode) => IfrsVariablesByImportIdentity.TryGetValue(dataNode, out var variableCollection) ? variableCollection : Enumerable.Empty<IfrsVariable>();",
                 "\n    ",
-                "\n    public double[] GetValues(ImportIdentity id, Func<RawVariable, bool> whereClause) => GetRawVariables(id.DataNode).Where(v => (v.AocType, v.Novelty) == id.AocStep && whereClause(v)).Select(v => v?.Values ?? (double[])null).AggregateDoubleArray();",
-                "\n    public double GetValue(ImportIdentity id, Func<IfrsVariable, bool> whereClause, int projection = 0) => GetIfrsVariables(id.DataNode).Where(v => (v.AocType, v.Novelty) == id.AocStep && whereClause(v)).Select(v => v?.Values ?? (double[])null).AggregateDoubleArray().ElementAtOrDefault(projection);",
+                "\n    public double[] GetValues(ImportIdentity id, Func<RawVariable, bool> whereClause) => GetRawVariables(id.DataNode).Where(v => new AocStep(v.AocType, v.Novelty) == id.AocStep && whereClause(v)).Select(v => v?.Values ?? (double[])null).AggregateDoubleArray();",
```
Comment (dcolleoni) [inline ifrs17/Import/ImportStorage.ipynb]: Should we replace the creation of an AocStep with individual comparison of AocType and Novelty

File: ifrs17/Report/ReportMutableScopes.ipynb [hunk 5]
```diff
@@ -116,11 +116,12 @@
                 "\n    string Scenario { get; set; }  // TODO: Enable dropdown selection including All and Delta",
                 "\n    string Comparison { get; set; }  // TODO: only for scenario at the beginning, meant to enable general purpose comparisons ",
                 "\n    CurrencyType CurrencyType { get; set; }",
-                "\n    string Projection {get; set;}",
+                "\n    string Projection {get; set;} //NB input is P0 or P1 or P15...",
                 "\n    ",
                 "\n    private ProjectionConfiguration[] projectionConfigurations => GetStorage().ProjectionConfigurations.SortRelevantProjections(Month);",
                 "\n    private string lastValidProjection => Projection == null ? null ",
-                "\n                                            : \"P\"+projectionConfigurations.Select(x => x.SystemName.Split(\"P\")[1]).Where(x => int.Parse(x) <= int.Parse(Projection.Split(\"P\")[1])).Last();",
+                "\n        //: \"P\" + projectionConfigurations.Select(x => x.SystemName.Split(\"P\")[1]).Where(x => int.Parse(x) <= int.Parse(Projection.Split(\"P\")[1])).Last(); //Old Code",
```
Comment (dcolleoni) [inline ifrs17/Report/ReportMutableScopes.ipynb]: clean up

File: ifrs17/Test/TestData.ipynb [hunk 6]
```diff
@@ -99,22 +99,22 @@
             "source": [
                 "var estimateType = ",
                 "\n@\"@@EstimateType,,,,,,,,,,,",
-                "\nSystemName,DisplayName,Order,StructureType,InputSource,PeriodType,ExternalId0,ExternalId1,ExternalId2,ExternalId3,ExternalId4,",
+                "\nSystemName,DisplayName,Order,StructureType,InputSource,PeriodType,,,,,,,",
```
Comment (dcolleoni) [inline ifrs17/Test/TestData.ipynb]: Add back the ExternalIds

File: ifrs17-template/Constants/CalculationEngine.ipynb
Comment (dcolleoni) [inline ifrs17-template/Constants/CalculationEngine.ipynb:1]: reset

File: ifrs17/Import/6ImportScope-Compute.ipynb [hunk 7]
```diff
@@ -164,10 +166,92 @@
...
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "public interface ComputeIfrsVarsActualsWithProjection : IScope<string, ImportStorage>{",
+                "\n    private IEnumerable<ImportIdentity> identities => GetScope<GetIdentities>(Identity).Identities;",
+                "\n    ",
+                "\n    private IEnumerable<ImportIdentity> identitiesWithProjection => Enumerable.Range(0, GetScope<ProjectionCount>(Identity).Value)",
```
Comment (dcolleoni) [inline ifrs17/Import/6ImportScope-Compute.ipynb]: extract to an interface to avoid duplication

File: ifrs17/Import/6ImportScope-Compute.ipynb [hunk 8]
```diff
@@ -164,10 +166,92 @@
         {
             "cell_type": "code",
             "source": [
-                "public interface ComputeIfrsVarsOpenings : ActualToIfrsVariable, DeferrableToIfrsVariable, TmToIfrsVariable",
-                "\n{",
-                "\n    IEnumerable<IfrsVariable> CalculatedIfrsVariables => AdvanceActual.Concat(OverdueActual)",
-                "\n                                                                      .Concat(Deferrable).Concat(Csms).Concat(Loss);",
+                "public interface ProjectionCount : IScope<string, ImportStorage>{",
```
Comment (dcolleoni) [inline ifrs17/Import/6ImportScope-Compute.ipynb]: Not a scope but a function

File: ifrs17/Import/6ImportScope-Compute.ipynb [hunk 9]
```diff
@@ -164,10 +166,92 @@
...
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "public interface ComputeIfrsVarsCashflowsWithProjection : IScope<string, ImportStorage>{",
+                "\n    private IEnumerable<ImportIdentity> identities => GetScope<GetIdentities>(Identity).Identities;",
+                "\n    ",
+                "\n    private IEnumerable<ImportIdentity> identitiesWithProjection => Enumerable.Range(0, GetScope<ProjectionCount>(Identity).Value)",
+                "\n        .SelectMany(projectionPeriod => identities.Select(id => id with { ProjectionPeriod = projectionPeriod})); //temp?",
```
Comment (dcolleoni) [inline ifrs17/Import/6ImportScope-Compute.ipynb]: clean up the comment

File: ifrs17/Import/6ImportScope-Compute.ipynb [hunk 10]
```diff
@@ -141,11 +144,10 @@
...
             "source": [
-                "public interface ComputeIfrsVarsCashflowsWithProjection : IScope<ImportIdentity, ImportStorage>{",
-                "\n    IEnumerable<IfrsVariable> CalculatedIfrsVariables => Enumerable.Range(0,ProjectionCount)",
-                "\n                                                                    .SelectMany(x => GetScope<ComputeIfrsVarsCashflows>(Identity with { ProjectionPeriod = x} ).CalculatedIfrsVariables)",
-                "\n                                                                    .AggregateProjections()",
-                "\n                                                                    .Select(x => x with {Partition = GetStorage().TargetPartition});",
+                "public interface ComputeIfrsVarsOpenings : ActualToIfrsVariable, DeferrableToIfrsVariable, TmToIfrsVariable",
+                "\n{",
+                "\n    IEnumerable<IfrsVariable> CalculatedIfrsVariables => AdvanceActual.Concat(OverdueActual)",
+                "\n        .Concat(Deferrable).Concat(Csms).Concat(Loss);//.Concat(Revenue)?",
```
Comment (dcolleoni) [inline ifrs17/Import/6ImportScope-Compute.ipynb]: Clean up + add Revenue

File: ifrs17/Import/3ImportScope-Actuals.ipynb [hunk 11]
```diff
@@ -120,38 +107,70 @@
...
-                "\n                                               .WithApplicability<ActualEqualsExpected>(x => !x.GetStorage().IsSecondaryScope(x.Identity.Id.DataNode) && x.Identity.Id.AocType == AocTypes.CF && x.Identity.Id.Novelty == Novelties.C && x.Identity.Id.ProjectionPeriod > 0 ));",
-                "\n",
-                "\n    public double Value => GetStorage().GetValue(Identity.Id, Identity.AmountType, Identity.EstimateType, Identity.AccidentYear, Identity.Id.ProjectionPeriod); ",
+                "public interface WrittenActual : IScope<(ImportIdentity Id, string AmountType, string EstimateType, int? AccidentYear), ImportStorage>{",
+                "\n            static ApplicabilityBuilder ScopeApplicabilityBuilder(ApplicabilityBuilder builder) =>",
+                "\n        builder.ForScope<WrittenActual>(s => s",
+                "\n            .WithApplicability<ActualEmptyValue>(x => x.GetStorage().ImportFormat == ImportFormats.Actual && !x.GetStorage().IsSecondaryScope(x.Identity.Id.DataNode) && //are these conditions required",
```
Comment (dcolleoni) [inline ifrs17/Import/3ImportScope-Actuals.ipynb]: clean up comment

File: ifrs17/Import/3ImportScope-Actuals.ipynb [hunk 12]
```diff
@@ -97,21 +91,14 @@
...
-                "\n",
-                "\nwhere the value is also function of the [Estimate Type](../DataModel/DataStructure#EstimateType) and [Amount Type](../DataModel/DataStructure#AmountType), and the $\\text{Imported Actual}$ value is described [here]()."
+                "## Actuals",
+                "\nThe Actuals correspond to WrittenActuals values and have estimate type $A$.",
+                "\nThe only valid AoC Step is Cashflow (CF).",
+                "\nProjection of Actuals correspond to the expected release of the best estimate: ",
+                "\n $$",
+                "\n \\text{Actual}(p) = \\Sigma_{\\text{Novelty}} \\text{PresentValue}(\\text{CF}, \\text{Novelty}),",
+                "\n $$",
+                "\n where p defines a future projected period."
```
Comment (dcolleoni) [inline ifrs17/Import/3ImportScope-Actuals.ipynb]: Add documentation around how to provide actuals vs PCE

File: ifrs17/Import/3ImportScope-Actuals.ipynb [hunk 13]
```diff
@@ -120,38 +107,70 @@
...
-                "\n",
-                "\n    public double Value => GetStorage().GetValue(Identity.Id, Identity.AmountType, Identity.EstimateType, Identity.AccidentYear, Identity.Id.ProjectionPeriod); ",
+                "public interface WrittenActual : IScope<(ImportIdentity Id, string AmountType, string EstimateType, int? AccidentYear), ImportStorage>{",
+                "\n            static ApplicabilityBuilder ScopeApplicabilityBuilder(ApplicabilityBuilder builder) =>",
+                "\n        builder.ForScope<WrittenActual>(s => s",
+                "\n            .WithApplicability<ActualEmptyValue>(x => x.GetStorage().ImportFormat == ImportFormats.Actual && !x.GetStorage().IsSecondaryScope(x.Identity.Id.DataNode) && //are these conditions required",
+                "\n                x.Identity.Id.AocType != AocTypes.CF)",
+                "\n            .WithApplicability<ActualEqualsExpected>(x => !x.GetStorage().IsSecondaryScope(x.Identity.Id.DataNode) && x.Identity.Id.Novelty == Novelties.C && x.Identity.Id.ProjectionPeriod > 0)",
```
Comment (dcolleoni) [inline ifrs17/Import/3ImportScope-Actuals.ipynb]: In the case of Projection PCE should have priority over the expected.
+ Documentation

File: ifrs17/Import/3ImportScope-Actuals.ipynb [hunk 14]
```diff
@@ -175,17 +203,38 @@
...
+                "\n            .WithApplicability<AccrualProjectionFirstYear>(x => !x.GetStorage().IsSecondaryScope(x.Identity.Id.DataNode) && x.GetStorage().GetShift(x.Identity.Id.ProjectionPeriod) == MonthInAYear && // projections ReportingYear +1",
+                "\n                                                                (x.Identity.Id.AocType == AocTypes.BOP || x.Identity.Id.AocType == AocTypes.CF))",
+                "\n            .WithApplicability<AccrualEmptyValues>(x => x.GetStorage().GetShift(x.Identity.Id.ProjectionPeriod) == MonthInAYear && x.Identity.Id.AocType == AocTypes.WO)                                                                ",
+                "\n            .WithApplicability<AccrualProjectionWithinFirstYear>(x => !x.GetStorage().IsSecondaryScope(x.Identity.Id.DataNode) && x.Identity.Id.ProjectionPeriod > 0)",
+                "\n            );",
+                "\n//TODO WO first year projection gets a value while it should be 0",
```
Comment (dcolleoni) [inline ifrs17/Import/3ImportScope-Actuals.ipynb]: clean up

File: ifrs17/Import/3ImportScope-Actuals.ipynb [hunk 15]
```diff
@@ -160,13 +179,22 @@
...
+                "\n\\text{Accrual Actual}(\\rm{BOP}) + \\text{Accrual Actual}(\\rm{CF}) + \\text{Accrual Actual}(\\rm{WO}) ~, & \\text{if AoC Step's AoC Type = EOP} \\\\",
+                "\n\\text{Imported Actual} ~, & \\text{otherwise}",
+                "\n\\end{array}",
+                "\n\\right.",
+                "\n$$",
+                "\n",
+                "\nwhere the value is also function of the [Estimate Type](../DataModel/DataStructure#EstimateType) and [Amount Type](../DataModel/DataStructure#AmountType).",
+                "\n",
+                "\nProjected values of accrual actuals within the current reporting year (if any) correspond to the current period figures, in respect to the year-to-date view (YTD) and because no futher assumptions can be made on actuals paid outside the period. In the first year projection we assume that all accrual actuals are released through the cashflow step thus not contributing to the projected Financial Perfomance. This results in closing the first year projection with zero accrual actuals and subsequent projections of accrual actuals to be zero. "
```
Comment (dcolleoni) [inline ifrs17/Import/3ImportScope-Actuals.ipynb]: first year projection with zero accrual => first year projection with non zero accrual

File: ifrs17/Import/2ImportScope-PresentValue.ipynb [hunk 16]
```diff
@@ -315,31 +315,31 @@
...
-                "\n{     ",
+                "\npublic interface DiscountedCreditRiskCashflow : DiscountedCashflow{",
                 "\n    private string cdrBasis => Identity.AmountType == AmountTypes.CDR ? EconomicBases.C : EconomicBases.L;",
                 "\n    private double nonPerformanceRiskRate => GetStorage().GetNonPerformanceRiskRate(Identity.Id, cdrBasis);",
                 "\n        ",
                 "\n    double[] DiscountedCashflow.Values => Multiply(-1d, NominalValues.ComputeDiscountAndCumulateWithCreditDefaultRisk(MonthlyDiscounting, nonPerformanceRiskRate)); // we need to flip the sign to create a reserve view",
+                "\n}",
+                "\n",
+                "\npublic interface DiscountedCashflowNextYearsProjection : DiscountedCashflow{",
+                "\n    double[] DiscountedCashflow.Values => GetScope<DiscountedCashflow>((Identity.Id with {ProjectionPeriod = GetStorage().NextYearProjection}, Identity.AmountType, Identity.EstimateType, Identity.AccidentYear)).Values;",
```
Comment (dcolleoni) [inline ifrs17/Import/2ImportScope-PresentValue.ipynb]: Improve name => (FirstNextYearProjection)

File: ifrs17/Import/ImportStorage.ipynb (hunk 4 above)
Comment (dcolleoni) [inline ifrs17/Import/ImportStorage.ipynb]: Compare AocType and Novelty separately

File: ifrs17/Import/1ImportScope-Identities.ipynb [hunk 17]
```diff
@@ -184,10 +184,11 @@
                 "\n                    ? new []{AmountTypes.CDR} : new []{AmountTypes.CDRI, AmountTypes.CDR} ",
                 "\n                    : Enumerable.Empty<string>()).ToHashSet();",
                 "\n    ",
-                "\n    ",
                 "\n    IEnumerable<string> ActualAmountTypes => GetStorage().GetIfrsVariables(Identity)",
                 "\n        .Where(iv => GetStorage().EstimateTypesByImportFormat[ImportFormats.Actual].Contains(iv.EstimateType))",
                 "\n        .Select(x => x.AmountType).ToHashSet();",
+                "\n",
+                "\n    IEnumerable<string> AllImportedAmountTypes => BeAmountTypes.Union(ActualAmountTypes);",
```
Comment (dcolleoni) [inline ifrs17/Import/1ImportScope-Identities.ipynb]: ToHashSet()

File: ifrs17/Import/Importers.ipynb [hunk 18]
```diff
@@ -644,6 +643,8 @@
                 "\n    await workspace.UpdateAsync<IfrsVariable>(ivs.Where(x => storage.DefaultPartition != storage.TargetPartition || x.Values.Any(v => Math.Abs(v) >= Precision)));",
                 "\n    await workspace.CommitToAsync<IfrsVariable,PartitionByReportingNodeAndPeriod>(workspaceToCompute, storage.TargetPartition, snapshot : true, ",
                 "\n                filter : x => storage.EstimateTypesByImportFormat[args.ImportFormat].Contains(x.EstimateType) ",
+                "\n// only the estimate types related to the importformat are saved:",
```
Comment (dcolleoni) [inline ifrs17/Import/Importers.ipynb]: Clean up

**Review by dtrzesniak — APPROVED** [review]

File: ifrs17-template/Test/IfrsVariablesTest.ipynb [hunk 19]
```diff
@@ -110,13 +110,12 @@
...
-                "\n{",
+                "public double? ParseIfrsVariableValue(IDataRow row, string nameOfIfrsVariableValue){",
                 "\n    return (row.Table.Columns.Contains(nameOfIfrsVariableValue)",
-                "\n                        ? (double.TryParse(row[nameOfIfrsVariableValue]?.ToString(), AllowedNumberStyles, CultureInfo.InvariantCulture, out var value) ",
-                "\n                                ? value ",
-                "\n                                : throw new Exception(nameOfIfrsVariableValue + \" cannot be parsed.\"))",
-                "\n                        : default(double));",
+                "\n        ? (double.TryParse(row[nameOfIfrsVariableValue]?.ToString(), AllowedNumberStyles, CultureInfo.InvariantCulture, out var value)",
+                "\n                ? value",
+                "\n                : default(double?)) //throw new Exception(nameOfIfrsVariableValue + \" cannot be parsed.\"))",
```
Comment (nnikolopoulos) [inline ifrs17-template/Test/IfrsVariablesTest.ipynb]: Is the throw now obsolete? Should it be cleaned up?

File: ifrs17-template/Test/IfrsVariablesTest.ipynb [hunk 20]
```diff
@@ -372,7 +393,7 @@
         {
             "cell_type": "code",
             "source": [
-                "excludedDataNodes = new string[]{\"DTP1.1\"};",
+                "excludedDataNodes = null; // new string[]{\"DTP1.1\"};",
```
Comment (nnikolopoulos) [inline ifrs17-template/Test/IfrsVariablesTest.ipynb:396]: cleanup?

File: ifrs17/Import/6ImportScope-Compute.ipynb (hunk 7 above)
Comment (dcolleoni) [reply ifrs17/Import/6ImportScope-Compute.ipynb]: Even better, I have simplified the scopes and ended up with only 1 scope. No repetition and clearer flow.

File: ifrs17/Import/6ImportScope-Compute.ipynb (hunk 8 above)
Comment (dcolleoni) [reply ifrs17/Import/6ImportScope-Compute.ipynb]: done

File: ifrs17-template/Test/IfrsVariablesTest.ipynb (hunk 20 above)
Comment (dcolleoni) [reply ifrs17-template/Test/IfrsVariablesTest.ipynb:396]: I would leave it to serve as an example

**Pull request merged**